### DISCUSSION TITLE
Custom themes & rasterizer

### DISCUSSION THREAD

**Discussion by asimaranov** [emilk/egui#544] [Q&A]
Hi! I'm trying to port the library to my e-ink reader. There are two issues I faced with

* Components are too small on 1404 * 1872 screen and I couldn't find how to enlarge it.
* I haven't found any information on how themes work. I need to define my own custom theme for black-and-white screen

Feature request:
To output graphics to a device without gpu I made a simple rasterizer, that accepts `Vec<ClippedShape>` and draws them to framebuffer. I think it would be useful to add a native rasterizer to the crate to run it on devices without gpu
![IMAGE 2021-07-04 09:07:52](https://user-images.githubusercontent.com/25568730/124374869-504ffa80-dca7-11eb-8dc8-196c4068dff8.jpg)

**Comment by parasyte** [comment]
When you create the `egui::RawInput` for your backend, there is a field you can set called [`pixels_per_point`](https://docs.rs/egui/0.13.1/egui/struct.RawInput.html#structfield.pixels_per_point). You can use this value to scale the UI.

Themes are defined with [`egui::Visuals`](https://docs.rs/egui/0.13.1/egui/style/struct.Visuals.html) and [`Context::set_visuals()`](https://docs.rs/egui/0.13.1/egui/struct.Context.html#method.set_visuals) can update it. There is also a higher-level [`Style`](https://docs.rs/egui/0.13.1/egui/style/struct.Style.html) (which composes `Visuals`) that can be set with `Context::set_style()` instead.

**Reply by asimaranov** [reply]
Thanks for answer. I tried to set the ppp option, but it seems not to work. 
ppp=1
![image](https://user-images.githubusercontent.com/25568730/124477125-4f54c100-ddac-11eb-844e-f3a697d70f5a.png)
ppp=10
![image](https://user-images.githubusercontent.com/25568730/124477201-6abfcc00-ddac-11eb-92ad-4bb370060e1f.png)
ppp = 0.5
![image](https://user-images.githubusercontent.com/25568730/124477744-09e4c380-ddad-11eb-8022-0be19b8aa6f8.png)

This is visualized tesselation output triangles, but without textures. Ui seems not to be scaled.

**Reply by parasyte** [reply]
Did you also notice that the [`RawInput::screen_rect`](https://docs.rs/egui/0.13.1/egui/struct.RawInput.html#structfield.screen_rect) should be provided in points (not pixels)? Your rasterizer needs to scale this rect to the physical display size.

The `egui-glium` and `egui-web` backends use OpenGL to do the scaling. Here's `egui-glium`: https://github.com/emilk/egui/blob/faf104220be18cda878e9864bab3760caa4901d8/egui_glium/src/painter.rs#L152-L154 https://github.com/emilk/egui/blob/faf104220be18cda878e9864bab3760caa4901d8/egui_glium/src/painter.rs#L161-L164

And here's similar code in `egui-web`: https://github.com/emilk/egui/blob/faf104220be18cda878e9864bab3760caa4901d8/egui_web/src/webgl2.rs#L434-L443

In both cases, the physical frame buffer size (in pixels) is transformed into a logical frame buffer size (in points; not coincidentally this should match the `RawInput::screen_rect` size) and the shader will use it to scale all vertex positions to the `[-1, 1]` range (normalized device coordinates). There's really no magic with the `pixels_per_point` value, other than GPUs being very efficient at scaling.

**Reply by asimaranov** [reply]
But isn't there a way to just enlarge all the component sizes? Maybe using custom theme? Out-the-box component sizes are constant?

**Reply by emilk** [reply]
You can make every font larger, and every thing in https://docs.rs/egui/0.13.1/egui/style/struct.Spacing.html bigger, but that's a bad way to go about it.

The better way it to follow @parasyte's advice. Instead of telling egui "I have a 1404x1872 screen with pixels_per_point=1" you say "I have a 702x936 screen with pixels_per_point=2".

**Reply by emilk** [reply]
Also: it would be VERY cool to see egui on a e-ink device 😍

**Reply by asimaranov** [reply]
![telegram-cloud-photo-size-2-5415653039987078068-y](https://user-images.githubusercontent.com/25568730/124670410-a2b63480-debc-11eb-963f-b9c8ac6f3c30.jpg)
Thanks, now it looks better
